# Buổi 03 & 04 - Bài 02 - Bài Tập Tổng Hợp

- MSSV: 25410291
- Họ Tên: Đinh Xuân Sâm
- GDrive: [IE313-b03-Bai02.ipynb](https://colab.research.google.com/drive/1N-pNcXCixloVnUNra3I9uUEtx3ix6WB7?usp=sharing)


Yêu cầu:

Sử dụng bộ dữ liệu thô ở chương 1 (link trong slide). Sau đó, tiến hành xử lý theo quy trình sau:

1. Cột `price` => `dropna` => dòng.
2. Cột `horsepower` => `replace` => `mean`.
3. Cột `normalized-losses` => `replace` => `mode`.
4. Bước cuối cùng, `dropna` dòng cho toàn bộ dữ liệu.
5. Quan sát giá trị thực tế, sau đó chuyển kiểu phù hợp cho các cột còn lại.
6. Chọn ra các cột số, chuẩn hóa thành `zscore`.

Nộp file code dạng `ipynb`.

## 0. Nạp Dữ Liệu

In [194]:
# Đạt biến chứa url/tên và nếu dùng gdrive cho dataset
dataset_url = "https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data"
dataset_file = "imports-85.csv"
dataset_file_gdrive = "/content/drive/MyDrive/uit/courses/IE313/06_Dataset/imports-85.csv"

In [195]:
# Tải dataset
import requests

response = requests.get(dataset_url)

with open(dataset_file, "wb") as file:
    file.write(response.content)

print("Download complete!")

Download complete!


In [196]:
# dataset headers
headers = [
"symboling",
"normalized-losses",
"make",
"fuel-type",
"aspiration",
"num-of-doors",
"body-style",
"drive-wheels",
"engine-location",
"wheel-base",
"length",
"width",
"height",
"curb-weight",
"engine-type",
"num-of-cylinders",
"engine-size",
"fuel-system",
"bore","stroke",
"compression-ratio",
"horsepower",
"peak-rpm",
"city-mpg",
"highway-mpg",
"price"]

In [197]:
# import panda
import pandas as pd

In [198]:
# đọc file dataset với headers chỉ định
df = pd.read_csv(dataset_file, names=headers)

In [199]:
# df quick look
df

,symboling,normalized-losses,make,fuel-type,aspiration,num-of-doors,body-style,drive-wheels,engine-location,wheel-base,...,engine-size,fuel-system,bore,stroke,compression-ratio,horsepower,peak-rpm,city-mpg,highway-mpg,price
0,3,?,alfa-romero,gas,std,two,convertible,rwd,front,88.6,...,130,mpfi,3.47,2.68,9.0,111,5000,21,27,13495
1,3,?,alfa-romero,gas,std,two,convertible,rwd,front,88.6,...,130,mpfi,3.47,2.68,9.0,111,5000,21,27,16500
2,1,?,alfa-romero,gas,std,two,hatchback,rwd,front,94.5,...,152,mpfi,2.68,3.47,9.0,154,5000,19,26,16500
3,2,164,audi,gas,std,four,sedan,fwd,front,99.8,...,109,mpfi,3.19,3.40,10.0,102,5500,24,30,13950
4,2,164,audi,gas,std,four,sedan,4wd,front,99.4,...,136,mpfi,3.19,3.40,8.0,115,5500,18,22,17450
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
200,-1,95,volvo,gas,std,four,sedan,rwd,front,109.1,...,141,mpfi,3.78,3.15,9.5,114,5400,23,28,16845
201,-1,95,volvo,gas,turbo,four,sedan,rwd,front,109.1,...,141,mpfi,3.78,3.15,8.7,160,5300,19,25,19045
202,-1,95,volvo,gas,std,four,sedan,rwd,front,109.1,...,173,mpfi,3.58,2.87,8.8,134,5500,18,23,21485
203,-1,95,volvo,diesel,turbo,four,sedan,rwd,front,109.1,...,145,idi,3.01,3.40,23.0,106,4800,26,27,22470


## 1. Cột `price` => `dropna` => dòng.

In [200]:
# Kích thước của df hiện tại
df.shape

(205, 26)

In [201]:
# Thông tin của df
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          205 non-null    int64  
 1   normalized-losses  205 non-null    object 
 2   make               205 non-null    object 
 3   fuel-type          205 non-null    object 
 4   aspiration         205 non-null    object 
 5   num-of-doors       205 non-null    object 
 6   body-style         205 non-null    object 
 7   drive-wheels       205 non-null    object 
 8   engine-location    205 non-null    object 
 9   wheel-base         205 non-null    float64
 10  length             205 non-null    float64
 11  width              205 non-null    float64
 12  height             205 non-null    float64
 13  curb-weight        205 non-null    int64  
 14  engine-type        205 non-null    object 
 15  num-of-cylinders   205 non-null    object 
 16  engine-size        205 non

In [202]:
# Kích thước của price
df['price'].shape

(205,)

In [203]:
# import numpy
import numpy as np

In [204]:
# Thay thế giá trị khuyết tự nhiên thành Numpy N/A và kiểm tra khuyết
df.replace(to_replace='?', value=np.nan, inplace=False).info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          205 non-null    int64  
 1   normalized-losses  164 non-null    object 
 2   make               205 non-null    object 
 3   fuel-type          205 non-null    object 
 4   aspiration         205 non-null    object 
 5   num-of-doors       203 non-null    object 
 6   body-style         205 non-null    object 
 7   drive-wheels       205 non-null    object 
 8   engine-location    205 non-null    object 
 9   wheel-base         205 non-null    float64
 10  length             205 non-null    float64
 11  width              205 non-null    float64
 12  height             205 non-null    float64
 13  curb-weight        205 non-null    int64  
 14  engine-type        205 non-null    object 
 15  num-of-cylinders   205 non-null    object 
 16  engine-size        205 non

### Nhận xét

- `price` có 201 dòng `non-null`, có nghĩa có 4 dòng N/A cần được loại bỏ

In [205]:
# Thay thế giá trị khuyết tự nhiên thành Numpy N/A, inplace
df.replace(to_replace='?', value=np.nan, inplace=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          205 non-null    int64  
 1   normalized-losses  164 non-null    object 
 2   make               205 non-null    object 
 3   fuel-type          205 non-null    object 
 4   aspiration         205 non-null    object 
 5   num-of-doors       203 non-null    object 
 6   body-style         205 non-null    object 
 7   drive-wheels       205 non-null    object 
 8   engine-location    205 non-null    object 
 9   wheel-base         205 non-null    float64
 10  length             205 non-null    float64
 11  width              205 non-null    float64
 12  height             205 non-null    float64
 13  curb-weight        205 non-null    int64  
 14  engine-type        205 non-null    object 
 15  num-of-cylinders   205 non-null    object 
 16  engine-size        205 non

In [206]:
# Kích thước của price sau khi điền N/A
df['price'].shape

(205,)

In [207]:
# Kiểm thử trước khi dropna
df.dropna(axis=0, subset=['price'], inplace=False).info()

<class 'pandas.core.frame.DataFrame'>
Index: 201 entries, 0 to 204
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          201 non-null    int64  
 1   normalized-losses  164 non-null    object 
 2   make               201 non-null    object 
 3   fuel-type          201 non-null    object 
 4   aspiration         201 non-null    object 
 5   num-of-doors       199 non-null    object 
 6   body-style         201 non-null    object 
 7   drive-wheels       201 non-null    object 
 8   engine-location    201 non-null    object 
 9   wheel-base         201 non-null    float64
 10  length             201 non-null    float64
 11  width              201 non-null    float64
 12  height             201 non-null    float64
 13  curb-weight        201 non-null    int64  
 14  engine-type        201 non-null    object 
 15  num-of-cylinders   201 non-null    object 
 16  engine-size        201 non-null

In [208]:
# Thực hiện dropna và kiểm tra lại kích thước của price
df.dropna(axis=0, subset=['price'], inplace=True)
df['price'].info()

<class 'pandas.core.series.Series'>
Index: 201 entries, 0 to 204
Series name: price
Non-Null Count  Dtype 
--------------  ----- 
201 non-null    object
dtypes: object(1)
memory usage: 3.1+ KB


In [209]:
# Kích thước của price sau khi dropna
df['price'].shape

(201,)

In [210]:
# Kiểm tra lại kích thước của df
df.shape

(201, 26)

### Kết Quả

- `price` đã được `dropna`.
- Số lượng dòng GIẢM từ 205 xuống 201 (tương ứng số dòng `non-null` đã đề cập trước đó).
- Số dòng của `df` GIẢM tương ứng vì dòng sẽ được xóa tương ứng.

## 2. Cột `horsepower` => `replace` => `mean`.

### Đánh Giá

- `horsepower` hiện đang có kiểu là `object`.
- Cần chuyển kiểu dữ liệu về kiểu số (`float`/`int`) để có thể tính `mean`.

In [211]:
# Lấy sample để đối chiếu kiểu dữ liệu
df.sample()

,symboling,normalized-losses,make,fuel-type,aspiration,num-of-doors,body-style,drive-wheels,engine-location,wheel-base,...,engine-size,fuel-system,bore,stroke,compression-ratio,horsepower,peak-rpm,city-mpg,highway-mpg,price
114,0,NaN,peugot,diesel,turbo,four,wagon,rwd,front,114.2,...,152,idi,3.70,3.52,21.0,95,4150,25,25,17075


In [212]:
# Kiểu dữ liệu hiện tại
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 201 entries, 0 to 204
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          201 non-null    int64  
 1   normalized-losses  164 non-null    object 
 2   make               201 non-null    object 
 3   fuel-type          201 non-null    object 
 4   aspiration         201 non-null    object 
 5   num-of-doors       199 non-null    object 
 6   body-style         201 non-null    object 
 7   drive-wheels       201 non-null    object 
 8   engine-location    201 non-null    object 
 9   wheel-base         201 non-null    float64
 10  length             201 non-null    float64
 11  width              201 non-null    float64
 12  height             201 non-null    float64
 13  curb-weight        201 non-null    int64  
 14  engine-type        201 non-null    object 
 15  num-of-cylinders   201 non-null    object 
 16  engine-size        201 non-null

Các cột cần đổi kiểu dữ liệu:

```
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          205 non-null    int64  
 1   normalized-losses  205 non-null    object -> float
 2   make               205 non-null    object
 3   fuel-type          205 non-null    object
 4   aspiration         205 non-null    object
 5   num-of-doors       205 non-null    object
 6   body-style         205 non-null    object
 7   drive-wheels       205 non-null    object
 8   engine-location    205 non-null    object
 9   wheel-base         205 non-null    float64
 10  length             205 non-null    float64
 11  width              205 non-null    float64
 12  height             205 non-null    float64
 13  curb-weight        205 non-null    int64  
 14  engine-type        205 non-null    object
 15  num-of-cylinders   205 non-null    object -> int
 16  engine-size        205 non-null    int64  
 17  fuel-system        205 non-null    object
 18  bore               205 non-null    object -> float
 19  stroke             205 non-null    object -> float
 20  compression-ratio  205 non-null    float64
 21  horsepower         205 non-null    object -> float
 22  peak-rpm           205 non-null    object -> int
 23  city-mpg           205 non-null    int64  
 24  highway-mpg        205 non-null    int64  
 25  price              205 non-null    object -> float
```

In [213]:
# Đổi kiểu dữ liệu để phản ánh đúng kiểu của giá trị
df['horsepower'] = df['horsepower'].astype('float')
df['horsepower'].info()

<class 'pandas.core.series.Series'>
Index: 201 entries, 0 to 204
Series name: horsepower
Non-Null Count  Dtype  
--------------  -----  
199 non-null    float64
dtypes: float64(1)
memory usage: 3.1 KB


In [214]:
# Tính mean của horsepower
mean_horsepower = df['horsepower'].mean()
mean_horsepower

np.float64(103.39698492462311)

In [215]:
# Kiểm tra trước kết quả của phép thay thế NA với mean
df['horsepower'].replace(to_replace=np.nan, value=mean_horsepower, inplace=False).info()

<class 'pandas.core.series.Series'>
Index: 201 entries, 0 to 204
Series name: horsepower
Non-Null Count  Dtype  
--------------  -----  
201 non-null    float64
dtypes: float64(1)
memory usage: 3.1 KB


In [216]:
# Thực hiện phép thay thế NA với mean
df['horsepower'].replace(to_replace=np.nan, value=mean_horsepower, inplace=True)
df['horsepower'].info()

<class 'pandas.core.series.Series'>
Index: 201 entries, 0 to 204
Series name: horsepower
Non-Null Count  Dtype  
--------------  -----  
201 non-null    float64
dtypes: float64(1)
memory usage: 3.1 KB


/tmp/ipykernel_1296/889122992.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['horsepower'].replace(to_replace=np.nan, value=mean_horsepower, inplace=True)


In [217]:
# Kiểm tra lại
df['horsepower']

,horsepower
0,111.0
1,111.0
2,154.0
3,102.0
4,115.0
...,...
200,114.0
201,160.0
202,134.0
203,106.0


### Kết Quả

- `horsepower` đã được đổi kiểu dữ liệu từ `object` thành `float`.
- Các giá trị `NaN` trong `horsepower` đã được thay thế bằng `mean` của `horsepower`.
- Tất cả 201 dòng của `horsepower` hiện đã không chứa bắt kỳ `NaN` nào nữa.

## 3. Cột `normalized-losses` => `replace` => `mode`.

In [218]:
# Xem thông tin của `normalized-lossess`
df['normalized-losses'].info()

<class 'pandas.core.series.Series'>
Index: 201 entries, 0 to 204
Series name: normalized-losses
Non-Null Count  Dtype 
--------------  ----- 
164 non-null    object
dtypes: object(1)
memory usage: 3.1+ KB


In [219]:
# Show N/A
df[df['normalized-losses'].isna()]['normalized-losses']

,normalized-losses
0,NaN
1,NaN
2,NaN
5,NaN
7,NaN
14,NaN
15,NaN
16,NaN
17,NaN
43,NaN


### Đánh Giá

- Tổng 201 dòng.
- Chỉ có 164 dòng có dữ liệu (non-null).
- Kiểu dữ liệu là `object`.

### Kế Hoạch

- Chuyển dữ liệu sang dạng số để có thể tính `mode`.
- Dùng `replace` để thay thế các giá trị null/NaN với `mode`.
- Kiểm tra lại kết quả.

In [220]:
# Convert data type
df['normalized-losses'] = df['normalized-losses'].astype('float')
df['normalized-losses'].info()

<class 'pandas.core.series.Series'>
Index: 201 entries, 0 to 204
Series name: normalized-losses
Non-Null Count  Dtype  
--------------  -----  
164 non-null    float64
dtypes: float64(1)
memory usage: 3.1 KB


In [221]:
# Calculate mode, get the very first value only
mode_normalized_losses = df['normalized-losses'].mode()[0]
mode_normalized_losses

np.float64(161.0)

In [222]:
# dry-run/preview the replace method
df['normalized-losses'].replace(to_replace=np.nan, value=mode_normalized_losses, inplace=False).info()

<class 'pandas.core.series.Series'>
Index: 201 entries, 0 to 204
Series name: normalized-losses
Non-Null Count  Dtype  
--------------  -----  
201 non-null    float64
dtypes: float64(1)
memory usage: 3.1 KB


In [223]:
# Apply the replace method
df['normalized-losses'].replace(to_replace=np.nan, value=mode_normalized_losses, inplace=True)
df['normalized-losses'].info()

<class 'pandas.core.series.Series'>
Index: 201 entries, 0 to 204
Series name: normalized-losses
Non-Null Count  Dtype  
--------------  -----  
201 non-null    float64
dtypes: float64(1)
memory usage: 3.1 KB


/tmp/ipykernel_1296/1817815612.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['normalized-losses'].replace(to_replace=np.nan, value=mode_normalized_losses, inplace=True)


In [224]:
# Show N/A
df[df['normalized-losses'].isna()]['normalized-losses']

,normalized-losses


In [225]:
# Preview
df['normalized-losses']

,normalized-losses
0,161.0
1,161.0
2,161.0
3,164.0
4,164.0
...,...
200,95.0
201,95.0
202,95.0
203,95.0


### Kết Quả

- `normalized-lossess` đã được thay thế các giá trị null/NaN với mode.
- Hiện tại không còn tồn tại dòng NaN nào trong `normalized-lossess` nữa.

## 4. Bước cuối cùng, `dropna` dòng cho toàn bộ dữ liệu.

### Kế Hoạch

- Xác định kích thước hiện tại của `df`.
- Kiểm tra số lượng hoặc/và xem trước các dòng/cột có NaN trong `df`.
- Thực hiện `dropna`.
- Kiểm tra lại kích thước của `df` và các giá trị NaN nếu còn tồn tại.

In [226]:
# df shape
df.shape

(201, 26)

In [227]:
# các dòng có NaN (ở bất kỳ cột nào)
df[df.isna().any(axis=1)]

,symboling,normalized-losses,make,fuel-type,aspiration,num-of-doors,body-style,drive-wheels,engine-location,wheel-base,...,engine-size,fuel-system,bore,stroke,compression-ratio,horsepower,peak-rpm,city-mpg,highway-mpg,price
27,1,148.0,dodge,gas,turbo,NaN,sedan,fwd,front,93.7,...,98,mpfi,3.03,3.39,7.6,102.000000,5500,24,30,8558
55,3,150.0,mazda,gas,std,two,hatchback,rwd,front,95.3,...,70,4bbl,NaN,NaN,9.4,101.000000,6000,17,23,10945
56,3,150.0,mazda,gas,std,two,hatchback,rwd,front,95.3,...,70,4bbl,NaN,NaN,9.4,101.000000,6000,17,23,11845
57,3,150.0,mazda,gas,std,two,hatchback,rwd,front,95.3,...,70,4bbl,NaN,NaN,9.4,101.000000,6000,17,23,13645
58,3,150.0,mazda,gas,std,two,hatchback,rwd,front,95.3,...,80,mpfi,NaN,NaN,9.4,135.000000,6000,16,23,15645
63,0,161.0,mazda,diesel,std,NaN,sedan,fwd,front,98.8,...,122,idi,3.39,3.39,22.7,64.000000,4650,36,42,10795
130,0,161.0,renault,gas,std,four,wagon,fwd,front,96.1,...,132,mpfi,3.46,3.90,8.7,103.396985,NaN,23,31,9295
131,2,161.0,renault,gas,std,two,hatchback,fwd,front,96.1,...,132,mpfi,3.46,3.90,8.7,103.396985,NaN,23,31,9895


In [228]:
# drop na for all the dataframe: dry-run
df.dropna(axis=0, subset=df.columns, inplace=False).info()

<class 'pandas.core.frame.DataFrame'>
Index: 193 entries, 0 to 204
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          193 non-null    int64  
 1   normalized-losses  193 non-null    float64
 2   make               193 non-null    object 
 3   fuel-type          193 non-null    object 
 4   aspiration         193 non-null    object 
 5   num-of-doors       193 non-null    object 
 6   body-style         193 non-null    object 
 7   drive-wheels       193 non-null    object 
 8   engine-location    193 non-null    object 
 9   wheel-base         193 non-null    float64
 10  length             193 non-null    float64
 11  width              193 non-null    float64
 12  height             193 non-null    float64
 13  curb-weight        193 non-null    int64  
 14  engine-type        193 non-null    object 
 15  num-of-cylinders   193 non-null    object 
 16  engine-size        193 non-null

In [229]:
# drop na for all the dataframe: execute
df.dropna(axis=0, subset=df.columns, inplace=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 193 entries, 0 to 204
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          193 non-null    int64  
 1   normalized-losses  193 non-null    float64
 2   make               193 non-null    object 
 3   fuel-type          193 non-null    object 
 4   aspiration         193 non-null    object 
 5   num-of-doors       193 non-null    object 
 6   body-style         193 non-null    object 
 7   drive-wheels       193 non-null    object 
 8   engine-location    193 non-null    object 
 9   wheel-base         193 non-null    float64
 10  length             193 non-null    float64
 11  width              193 non-null    float64
 12  height             193 non-null    float64
 13  curb-weight        193 non-null    int64  
 14  engine-type        193 non-null    object 
 15  num-of-cylinders   193 non-null    object 
 16  engine-size        193 non-null

In [230]:
# Kiểm tra lại các dòng có NaN (ở bất kỳ cột nào)
df[df.isna().any(axis=1)]

,symboling,normalized-losses,make,fuel-type,aspiration,num-of-doors,body-style,drive-wheels,engine-location,wheel-base,...,engine-size,fuel-system,bore,stroke,compression-ratio,horsepower,peak-rpm,city-mpg,highway-mpg,price


In [231]:
# Kiểm lại shape
df.shape

(193, 26)

### Kết Quả

- `df` hiện không tồn tại giá trị NaN nào nữa.
- Số dòng giảm xuống tương ứng với số dòng đã tồn tại NaN trước đó (201 - 8 = 193).


## 5. Chuyển kiểu phù hợp cho các cột còn lại.

In [232]:
# df info
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 193 entries, 0 to 204
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          193 non-null    int64  
 1   normalized-losses  193 non-null    float64
 2   make               193 non-null    object 
 3   fuel-type          193 non-null    object 
 4   aspiration         193 non-null    object 
 5   num-of-doors       193 non-null    object 
 6   body-style         193 non-null    object 
 7   drive-wheels       193 non-null    object 
 8   engine-location    193 non-null    object 
 9   wheel-base         193 non-null    float64
 10  length             193 non-null    float64
 11  width              193 non-null    float64
 12  height             193 non-null    float64
 13  curb-weight        193 non-null    int64  
 14  engine-type        193 non-null    object 
 15  num-of-cylinders   193 non-null    object 
 16  engine-size        193 non-null

In [233]:
# sample
df.sample()

,symboling,normalized-losses,make,fuel-type,aspiration,num-of-doors,body-style,drive-wheels,engine-location,wheel-base,...,engine-size,fuel-system,bore,stroke,compression-ratio,horsepower,peak-rpm,city-mpg,highway-mpg,price
186,2,94.0,volkswagen,gas,std,four,sedan,fwd,front,97.3,...,109,mpfi,3.19,3.40,9.0,85.0,5250,27,34,8495


### Nhận Xét

Dựa trên `df.info()` và `.sample()` ở tên, chúng ta xác định các cột cần đổi như dưới đây.

- Lý do: Nhận đúng dạng kiểu số nếu đang nhận là object.

Các cột cần đổi:

```
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  

 18  bore               193 non-null    object -> float
 19  stroke             193 non-null    object -> float

 22  peak-rpm           193 non-null    object -> int

 25  price              193 non-null    object -> float
```

In [234]:
# Kiểm lại giá trị thực tế của dữ liệu
df.loc[12, ['bore', 'stroke', 'peak-rpm', 'price']]

,12
bore,3.31
stroke,3.19
peak-rpm,4250
price,20970


In [235]:
# Kiểm tra kiểu dữ liệu trước khi thực hiện
df[['bore', 'stroke', 'peak-rpm', 'price']].info()

<class 'pandas.core.frame.DataFrame'>
Index: 193 entries, 0 to 204
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   bore      193 non-null    object
 1   stroke    193 non-null    object
 2   peak-rpm  193 non-null    object
 3   price     193 non-null    object
dtypes: object(4)
memory usage: 11.6+ KB


In [236]:
# Đổi kiểu dữ liệu
# Dù price trông rất giống int nhưng sẽ dùng float từ thuộc tính tự nhiên của nó
df['bore'] = df['bore'].astype('float')
df['stroke'] = df['stroke'].astype('float')
df['peak-rpm'] = df['peak-rpm'].astype('int')
df['price'] = df['price'].astype('float')
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 193 entries, 0 to 204
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          193 non-null    int64  
 1   normalized-losses  193 non-null    float64
 2   make               193 non-null    object 
 3   fuel-type          193 non-null    object 
 4   aspiration         193 non-null    object 
 5   num-of-doors       193 non-null    object 
 6   body-style         193 non-null    object 
 7   drive-wheels       193 non-null    object 
 8   engine-location    193 non-null    object 
 9   wheel-base         193 non-null    float64
 10  length             193 non-null    float64
 11  width              193 non-null    float64
 12  height             193 non-null    float64
 13  curb-weight        193 non-null    int64  
 14  engine-type        193 non-null    object 
 15  num-of-cylinders   193 non-null    object 
 16  engine-size        193 non-null

In [237]:
# Kiểm lại kiểu dữ liệu
df[['bore', 'stroke', 'peak-rpm', 'price']].info()

<class 'pandas.core.frame.DataFrame'>
Index: 193 entries, 0 to 204
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   bore      193 non-null    float64
 1   stroke    193 non-null    float64
 2   peak-rpm  193 non-null    int64  
 3   price     193 non-null    float64
dtypes: float64(3), int64(1)
memory usage: 11.6 KB


### Kết Quả

- Các cột đã có kiểu dữ liệu phù hợp với giá trị thực tế.
- Cụ thể các cột đã được cập nhật là: 'bore', 'stroke', 'peak-rpm', 'price'.

## 6. Chọn ra các cột số, chuẩn hóa thành `zscore`.

### Nhận Xét

- `df` có một số lượng lớn các cột có dạng số (numeric), bao gồm `int` và `float`.

In [238]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 193 entries, 0 to 204
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          193 non-null    int64  
 1   normalized-losses  193 non-null    float64
 2   make               193 non-null    object 
 3   fuel-type          193 non-null    object 
 4   aspiration         193 non-null    object 
 5   num-of-doors       193 non-null    object 
 6   body-style         193 non-null    object 
 7   drive-wheels       193 non-null    object 
 8   engine-location    193 non-null    object 
 9   wheel-base         193 non-null    float64
 10  length             193 non-null    float64
 11  width              193 non-null    float64
 12  height             193 non-null    float64
 13  curb-weight        193 non-null    int64  
 14  engine-type        193 non-null    object 
 15  num-of-cylinders   193 non-null    object 
 16  engine-size        193 non-null

In [239]:
df.dtypes

,0
symboling,int64
normalized-losses,float64
make,object
fuel-type,object
aspiration,object
num-of-doors,object
body-style,object
drive-wheels,object
engine-location,object
wheel-base,float64


### Kế Hoạch

- Lọc ra những cột có `dtype` là kiểu số.
- Sử dụng `sklearn.preprocessing.StandardScaler` để `transform` từng cột tương ứng.

In [240]:
# Chọn ra những cột có dạng số
df.select_dtypes(include=['int64', 'float64']).info()

<class 'pandas.core.frame.DataFrame'>
Index: 193 entries, 0 to 204
Data columns (total 16 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          193 non-null    int64  
 1   normalized-losses  193 non-null    float64
 2   wheel-base         193 non-null    float64
 3   length             193 non-null    float64
 4   width              193 non-null    float64
 5   height             193 non-null    float64
 6   curb-weight        193 non-null    int64  
 7   engine-size        193 non-null    int64  
 8   bore               193 non-null    float64
 9   stroke             193 non-null    float64
 10  compression-ratio  193 non-null    float64
 11  horsepower         193 non-null    float64
 12  peak-rpm           193 non-null    int64  
 13  city-mpg           193 non-null    int64  
 14  highway-mpg        193 non-null    int64  
 15  price              193 non-null    float64
dtypes: float64(10), int64(6)
memory

In [241]:
# Lấy ra danh sách các cột này
columns_list = df.select_dtypes(include=['int64', 'float64']).columns.to_list()
columns_list

['symboling',
 'normalized-losses',
 'wheel-base',
 'length',
 'width',
 'height',
 'curb-weight',
 'engine-size',
 'bore',
 'stroke',
 'compression-ratio',
 'horsepower',
 'peak-rpm',
 'city-mpg',
 'highway-mpg',
 'price']

In [242]:
# import StandardScaler
from sklearn.preprocessing import StandardScaler
ss = StandardScaler()

In [243]:
# Thực hiện transform zscore
for column in columns_list:
    df[column + '_zscore'] = ss.fit_transform(df[[column]])

In [244]:
# Kiểm lại df
df.shape

(193, 42)

In [245]:
# Xem sample
df.sample()

,symboling,normalized-losses,make,fuel-type,aspiration,num-of-doors,body-style,drive-wheels,engine-location,wheel-base,...,curb-weight_zscore,engine-size_zscore,bore_zscore,stroke_zscore,compression-ratio_zscore,horsepower_zscore,peak-rpm_zscore,city-mpg_zscore,highway-mpg_zscore,price_zscore
192,0,161.0,volkswagen,diesel,turbo,four,sedan,fwd,front,100.4,...,0.033297,-0.7503,-1.180154,0.480415,3.240689,-0.937146,-1.282927,1.204405,1.060773,0.069406


In [246]:
# Lấy ra các cột zscore
df[[ column + '_zscore' for column in columns_list]]

,symboling_zscore,normalized-losses_zscore,wheel-base_zscore,length_zscore,width_zscore,height_zscore,curb-weight_zscore,engine-size_zscore,bore_zscore,stroke_zscore,compression-ratio_zscore,horsepower_zscore,peak-rpm_zscore,city-mpg_zscore,highway-mpg_zscore,price_zscore
0,1.786850,0.921206,-1.682379,-0.444024,-0.841263,-2.122598,-0.025713,0.045215,0.513027,-1.808186,-0.288273,0.198569,-0.213359,-0.679053,-0.557058,0.026025
1,1.786850,0.921206,-1.682379,-0.444024,-0.841263,-2.122598,-0.025713,0.045215,0.513027,-1.808186,-0.288273,0.198569,-0.213359,-0.679053,-0.557058,0.398480
2,0.163970,0.921206,-0.720911,-0.251195,-0.184679,-0.615412,0.497764,0.575559,-2.394827,0.702918,-0.288273,1.334283,-0.213359,-0.992963,-0.704134,0.398480
3,0.975410,1.005348,0.142781,0.182672,0.143612,0.180047,-0.427362,-0.461021,-0.517605,0.480415,-0.036204,-0.039139,0.856208,-0.208189,-0.115832,0.082420
4,0.975410,1.005348,0.077596,0.182672,0.237410,0.180047,0.499668,0.189854,-0.517605,0.480415,-0.540341,0.304217,0.856208,-1.149918,-1.292436,0.516227
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
200,-1.458911,-0.929925,1.658315,1.162889,1.409880,0.682442,0.743322,0.310387,1.654084,-0.314238,-0.162238,0.277805,0.642295,-0.365144,-0.409983,0.441241
201,-1.458911,-0.929925,1.658315,1.162889,1.362982,0.682442,0.927967,0.310387,1.654084,-0.314238,-0.363893,1.492755,0.428381,-0.992963,-0.851209,0.713920
202,-1.458911,-0.929925,1.658315,1.162889,1.409880,0.682442,0.857535,1.081795,0.917918,-1.204249,-0.338686,0.806044,0.856208,-1.149918,-1.145360,1.016345
203,-1.458911,-0.929925,1.658315,1.162889,1.409880,0.682442,1.247763,0.406813,-1.180154,0.480415,3.240689,0.066509,-0.641186,0.105721,-0.557058,1.138431


### Kết Quả

- Có tổng số 16 cột dạng số.
- Tất cả các cột dạng số đã được tạo cột zscore, tương ứng 16 cột thêm mới.
- Kích thước của `df` hiện tại là (193, 42), với 42 = 26 cột đã có + 16 cột mới.

# Buổi 04 - Bài 02 - Bài Tập Tổng Hợp (Tiếp Tục Buổi 03)

Các yêu cầu:

1. Xử lý khuyết theo yêu cầu (đã hoàn thành ở bài tập của buổi 03)
2. Quan sát các cột nhận dạng là kiểu phân loại $⇒$ chuyển thành số (đã hoàn thành ở bài tập của buổi 03)
3. Phân loại (chữ sang số) sang số:
    - Integer encoding áp dụng cho `make`.
    - Ordinal encoding áp dụng cho `num-of-doors`.
    - One-hot encoding áp dụng còn lại.
4. Chuẩn hóa z-score (sẽ áp dụng cho các cột mới bổ sung từ bước 03).


## 3. Phân Loại

In [247]:
# df hiện tại
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 193 entries, 0 to 204
Data columns (total 42 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   symboling                 193 non-null    int64  
 1   normalized-losses         193 non-null    float64
 2   make                      193 non-null    object 
 3   fuel-type                 193 non-null    object 
 4   aspiration                193 non-null    object 
 5   num-of-doors              193 non-null    object 
 6   body-style                193 non-null    object 
 7   drive-wheels              193 non-null    object 
 8   engine-location           193 non-null    object 
 9   wheel-base                193 non-null    float64
 10  length                    193 non-null    float64
 11  width                     193 non-null    float64
 12  height                    193 non-null    float64
 13  curb-weight               193 non-null    int64  
 14  engine-type    

### Integer encoding cho make

Ý tưởng: lấy toàn bộ các giá trị trong `make`, sau đó map/gán tương ứng các giá trị số thành một `dict`, sau đó dùng `.replace()` để thực hiện việc chuyển đổi.

In [248]:
# xem thuộc tính của make, hiện tại dtype là object
df['make'].info()

<class 'pandas.core.series.Series'>
Index: 193 entries, 0 to 204
Series name: make
Non-Null Count  Dtype 
--------------  ----- 
193 non-null    object
dtypes: object(1)
memory usage: 7.1+ KB


In [249]:
# Các maker trong dataset
df['make'].unique()

array(['alfa-romero', 'audi', 'bmw', 'chevrolet', 'dodge', 'honda',
       'isuzu', 'jaguar', 'mazda', 'mercedes-benz', 'mercury',
       'mitsubishi', 'nissan', 'peugot', 'plymouth', 'porsche', 'saab',
       'subaru', 'toyota', 'volkswagen', 'volvo'], dtype=object)

In [250]:
# mỗi maker được map vào một số tương ứng
makers_count = 0
makers = {}

for maker in df['make'].unique():
    makers_count += 1
    makers[maker] = makers_count

makers

{'alfa-romero': 1,
 'audi': 2,
 'bmw': 3,
 'chevrolet': 4,
 'dodge': 5,
 'honda': 6,
 'isuzu': 7,
 'jaguar': 8,
 'mazda': 9,
 'mercedes-benz': 10,
 'mercury': 11,
 'mitsubishi': 12,
 'nissan': 13,
 'peugot': 14,
 'plymouth': 15,
 'porsche': 16,
 'saab': 17,
 'subaru': 18,
 'toyota': 19,
 'volkswagen': 20,
 'volvo': 21}

In [251]:
# thử trước khi replace
df['make'].replace(makers, inplace=False)

/tmp/ipykernel_1296/2621407543.py:2: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df['make'].replace(makers, inplace=False)


,make
0,1
1,1
2,1
3,2
4,2
...,...
200,21
201,21
202,21
203,21


In [252]:
# thực hiện replace
df['make'].replace(makers, inplace=True)

/tmp/ipykernel_1296/4257892115.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['make'].replace(makers, inplace=True)
/tmp/ipykernel_1296/4257892115.py:2: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df['make'].replace(makers, inplace=True)


In [253]:
# kiểm tra lại
df['make']

,make
0,1
1,1
2,1
3,2
4,2
...,...
200,21
201,21
202,21
203,21


In [254]:
# kiểm tra lại các loại
df['make'].unique()

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21])

In [255]:
# kiểm tra số lượng của mỗi giá trị tương ứng
df['make'].value_counts()

,count
make,
19,32
13,18
12,13
6,13
18,12
9,12
20,12
21,11
14,11


In [256]:
# xem thuộc tính của make, nếu cần thì sẽ đổi dtype tương ứng sang dạng số
df['make'].info()

<class 'pandas.core.series.Series'>
Index: 193 entries, 0 to 204
Series name: make
Non-Null Count  Dtype
--------------  -----
193 non-null    int64
dtypes: int64(1)
memory usage: 7.1 KB


### Ordinal encoding áp dụng cho `num-of-doors`.

In [257]:
# xem thông tin của series
df['num-of-doors'].info()

<class 'pandas.core.series.Series'>
Index: 193 entries, 0 to 204
Series name: num-of-doors
Non-Null Count  Dtype 
--------------  ----- 
193 non-null    object
dtypes: object(1)
memory usage: 7.1+ KB


In [258]:
# xác định các giá trị của series
df['num-of-doors'].unique()

array(['two', 'four'], dtype=object)

In [259]:
# cùng đếm xem mỗi loại có số lượng bao nhiêu
df['num-of-doors'].value_counts()

,count
num-of-doors,
four,112
two,81


In [260]:
# Chúng ta chỉ có 2 loại giá trị, two, four
# nên là tạo một dict để phân loại
doors = {
    'two': 2,
    'four': 4
}

In [261]:
# dùng replace để cập nhật/chuyển đổi
# phần này giống câu ở trên
# vẫn nên kiểm thử trước
df['num-of-doors'].replace(doors, inplace=False).unique()

/tmp/ipykernel_1296/1664137147.py:4: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df['num-of-doors'].replace(doors, inplace=False).unique()


array([2, 4])

In [262]:
# thực hiện
df['num-of-doors'].replace(doors, inplace=True)

/tmp/ipykernel_1296/265144393.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['num-of-doors'].replace(doors, inplace=True)
/tmp/ipykernel_1296/265144393.py:2: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df['num-of-doors'].replace(doors, inplace=True)


In [263]:
# kiểm lại
df['num-of-doors'].value_counts()

,count
num-of-doors,
4,112
2,81


In [264]:
# df hiện tại
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 193 entries, 0 to 204
Data columns (total 42 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   symboling                 193 non-null    int64  
 1   normalized-losses         193 non-null    float64
 2   make                      193 non-null    int64  
 3   fuel-type                 193 non-null    object 
 4   aspiration                193 non-null    object 
 5   num-of-doors              193 non-null    int64  
 6   body-style                193 non-null    object 
 7   drive-wheels              193 non-null    object 
 8   engine-location           193 non-null    object 
 9   wheel-base                193 non-null    float64
 10  length                    193 non-null    float64
 11  width                     193 non-null    float64
 12  height                    193 non-null    float64
 13  curb-weight               193 non-null    int64  
 14  engine-type    

### One-hot encoding áp dụng còn lại.

In [265]:
# xem nhanh df hiện tại
df.sample()

,symboling,normalized-losses,make,fuel-type,aspiration,num-of-doors,body-style,drive-wheels,engine-location,wheel-base,...,curb-weight_zscore,engine-size_zscore,bore_zscore,stroke_zscore,compression-ratio_zscore,horsepower_zscore,peak-rpm_zscore,city-mpg_zscore,highway-mpg_zscore,price_zscore
112,0,161.0,14,diesel,turbo,4,sedan,rwd,front,107.9,...,1.314388,0.575559,1.359618,0.861849,2.736552,-0.224023,-2.031625,0.419631,0.325395,0.448058


In [266]:
# xác định các cột có dtype là object và chưa xử lý
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 193 entries, 0 to 204
Data columns (total 42 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   symboling                 193 non-null    int64  
 1   normalized-losses         193 non-null    float64
 2   make                      193 non-null    int64  
 3   fuel-type                 193 non-null    object 
 4   aspiration                193 non-null    object 
 5   num-of-doors              193 non-null    int64  
 6   body-style                193 non-null    object 
 7   drive-wheels              193 non-null    object 
 8   engine-location           193 non-null    object 
 9   wheel-base                193 non-null    float64
 10  length                    193 non-null    float64
 11  width                     193 non-null    float64
 12  height                    193 non-null    float64
 13  curb-weight               193 non-null    int64  
 14  engine-type    

In [267]:
# lấy danh sách các column và dtype tương ứng
data_types = df.dtypes
data_types.info()

<class 'pandas.core.series.Series'>
Index: 42 entries, symboling to price_zscore
Series name: None
Non-Null Count  Dtype 
--------------  ----- 
42 non-null     object
dtypes: object(1)
memory usage: 1.7+ KB


In [268]:
# lọc ra các column có dtype là object
# đây chính là các column còn lại cần chuyển đổi
object_columns = data_types[data_types == 'object'].index.tolist()
object_columns

['fuel-type',
 'aspiration',
 'body-style',
 'drive-wheels',
 'engine-location',
 'engine-type',
 'num-of-cylinders',
 'fuel-system']

In [269]:
# thử lấy dummies của một column trong danh sách trên
# đảm bảo rằng dummies hợp lý để trả về true/false
pd.get_dummies(df[object_columns[0]], prefix=object_columns[0])

,fuel-type_diesel,fuel-type_gas
0,False,True
1,False,True
2,False,True
3,False,True
4,False,True
...,...,...
200,False,True
201,False,True
202,False,True
203,True,False


In [270]:
# xác định thử các giá trị của một column trong danh sách
df[object_columns[0]].value_counts()

,count
fuel-type,
gas,174
diesel,19


In [271]:
# với mỗi column trong danh sách, áp dụng one-hot encoding
for cot in object_columns:
    dummies = pd.get_dummies(df[cot], prefix=cot)
    df = pd.concat([df, dummies], axis=1)

# sau đó kiểm tra lại số cột/dtype tương ứng
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 193 entries, 0 to 204
Data columns (total 74 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   symboling                 193 non-null    int64  
 1   normalized-losses         193 non-null    float64
 2   make                      193 non-null    int64  
 3   fuel-type                 193 non-null    object 
 4   aspiration                193 non-null    object 
 5   num-of-doors              193 non-null    int64  
 6   body-style                193 non-null    object 
 7   drive-wheels              193 non-null    object 
 8   engine-location           193 non-null    object 
 9   wheel-base                193 non-null    float64
 10  length                    193 non-null    float64
 11  width                     193 non-null    float64
 12  height                    193 non-null    float64
 13  curb-weight               193 non-null    int64  
 14  engine-type    

In [272]:
# để làm sạch, chúng ta sẽ xóa bỏ các column đã được chuyển đổi
# nói cách khác, là xóa các columns có dtype object
df = df.drop(columns=object_columns)

# sau đó kiểm lại
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 193 entries, 0 to 204
Data columns (total 66 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   symboling                 193 non-null    int64  
 1   normalized-losses         193 non-null    float64
 2   make                      193 non-null    int64  
 3   num-of-doors              193 non-null    int64  
 4   wheel-base                193 non-null    float64
 5   length                    193 non-null    float64
 6   width                     193 non-null    float64
 7   height                    193 non-null    float64
 8   curb-weight               193 non-null    int64  
 9   engine-size               193 non-null    int64  
 10  bore                      193 non-null    float64
 11  stroke                    193 non-null    float64
 12  compression-ratio         193 non-null    float64
 13  horsepower                193 non-null    float64
 14  peak-rpm       

In [273]:
# có thể xem thử
display(df.head())

,symboling,normalized-losses,make,num-of-doors,wheel-base,length,width,height,curb-weight,engine-size,...,num-of-cylinders_six,num-of-cylinders_three,num-of-cylinders_twelve,fuel-system_1bbl,fuel-system_2bbl,fuel-system_idi,fuel-system_mfi,fuel-system_mpfi,fuel-system_spdi,fuel-system_spfi
0,3,161.0,1,2,88.6,168.8,64.1,48.8,2548,130,...,False,False,False,False,False,False,False,True,False,False
1,3,161.0,1,2,88.6,168.8,64.1,48.8,2548,130,...,False,False,False,False,False,False,False,True,False,False
2,1,161.0,1,2,94.5,171.2,65.5,52.4,2823,152,...,True,False,False,False,False,False,False,True,False,False
3,2,164.0,2,4,99.8,176.6,66.2,54.3,2337,109,...,False,False,False,False,False,False,False,True,False,False
4,2,164.0,2,4,99.4,176.6,66.4,54.3,2824,136,...,False,False,False,False,False,False,False,True,False,False


## Chuẩn Hóa z-score (cho các cột từ bước 3)

### Nhận Xét

- Một số lượng lớn các cột mới đã được tạo ra từ bước chuyển đổi ở trên.
- Kèm theo các cột từ bài tập 03.

### Kế Hoạch

Có các cách sau:

1. Xóa tất cả các cột z-score đã tạo ở buổi 03, và tạo mới zscore cho tất cả các cột.
2. Lọc ra các cột mới từ bước 03 và tạo cột zscore tương ứng.

Trong bài này sẽ làm theo cách 2, vì sát với đề bài của bài tập.

In [274]:
# xem lại các cột đã được tạo z-score ở bài của buổi 03
columns_list

['symboling',
 'normalized-losses',
 'wheel-base',
 'length',
 'width',
 'height',
 'curb-weight',
 'engine-size',
 'bore',
 'stroke',
 'compression-ratio',
 'horsepower',
 'peak-rpm',
 'city-mpg',
 'highway-mpg',
 'price']

In [275]:
# đây là các zscore columns đã được tạo từ bài tập buổi 03
zscore_columns = [column + '_zscore' for column in columns_list]
zscore_columns

['symboling_zscore',
 'normalized-losses_zscore',
 'wheel-base_zscore',
 'length_zscore',
 'width_zscore',
 'height_zscore',
 'curb-weight_zscore',
 'engine-size_zscore',
 'bore_zscore',
 'stroke_zscore',
 'compression-ratio_zscore',
 'horsepower_zscore',
 'peak-rpm_zscore',
 'city-mpg_zscore',
 'highway-mpg_zscore',
 'price_zscore']

In [276]:
# chọn ra các columns dạng số và bool
df.select_dtypes(include=['int64', 'float64', 'bool']).columns.to_list()

['symboling',
 'normalized-losses',
 'make',
 'num-of-doors',
 'wheel-base',
 'length',
 'width',
 'height',
 'curb-weight',
 'engine-size',
 'bore',
 'stroke',
 'compression-ratio',
 'horsepower',
 'peak-rpm',
 'city-mpg',
 'highway-mpg',
 'price',
 'symboling_zscore',
 'normalized-losses_zscore',
 'wheel-base_zscore',
 'length_zscore',
 'width_zscore',
 'height_zscore',
 'curb-weight_zscore',
 'engine-size_zscore',
 'bore_zscore',
 'stroke_zscore',
 'compression-ratio_zscore',
 'horsepower_zscore',
 'peak-rpm_zscore',
 'city-mpg_zscore',
 'highway-mpg_zscore',
 'price_zscore',
 'fuel-type_diesel',
 'fuel-type_gas',
 'aspiration_std',
 'aspiration_turbo',
 'body-style_convertible',
 'body-style_hardtop',
 'body-style_hatchback',
 'body-style_sedan',
 'body-style_wagon',
 'drive-wheels_4wd',
 'drive-wheels_fwd',
 'drive-wheels_rwd',
 'engine-location_front',
 'engine-location_rear',
 'engine-type_dohc',
 'engine-type_l',
 'engine-type_ohc',
 'engine-type_ohcf',
 'engine-type_ohcv',
 'n

In [277]:
# lấy ra những cột từ danh sách trên mà:
# 1. không thuộc các cột đã được tính zscore
# 2. không thuộc các cột là zscore
columns_to_zscore = [column for column in df.select_dtypes(include=['int64', 'float64', 'bool']).columns.to_list() if (column not in zscore_columns and column not in columns_list)]
columns_to_zscore

['make',
 'num-of-doors',
 'fuel-type_diesel',
 'fuel-type_gas',
 'aspiration_std',
 'aspiration_turbo',
 'body-style_convertible',
 'body-style_hardtop',
 'body-style_hatchback',
 'body-style_sedan',
 'body-style_wagon',
 'drive-wheels_4wd',
 'drive-wheels_fwd',
 'drive-wheels_rwd',
 'engine-location_front',
 'engine-location_rear',
 'engine-type_dohc',
 'engine-type_l',
 'engine-type_ohc',
 'engine-type_ohcf',
 'engine-type_ohcv',
 'num-of-cylinders_eight',
 'num-of-cylinders_five',
 'num-of-cylinders_four',
 'num-of-cylinders_six',
 'num-of-cylinders_three',
 'num-of-cylinders_twelve',
 'fuel-system_1bbl',
 'fuel-system_2bbl',
 'fuel-system_idi',
 'fuel-system_mfi',
 'fuel-system_mpfi',
 'fuel-system_spdi',
 'fuel-system_spfi']

In [278]:
df.shape

(193, 66)

In [279]:
# thực hiện tạo zscore cho các cột mới
# đồng thời thêm các cột zstore mới tạo vào danh sách các cột zstore trước đó

for column in columns_to_zscore:
    df[column + '_zscore'] = ss.fit_transform(df[[column]])
    zscore_columns.append(column + '_zscore')

In [280]:
# kiểm lại
df.shape

(193, 100)

In [281]:
df.sample()

,symboling,normalized-losses,make,num-of-doors,wheel-base,length,width,height,curb-weight,engine-size,...,num-of-cylinders_six_zscore,num-of-cylinders_three_zscore,num-of-cylinders_twelve_zscore,fuel-system_1bbl_zscore,fuel-system_2bbl_zscore,fuel-system_idi_zscore,fuel-system_mfi_zscore,fuel-system_mpfi_zscore,fuel-system_spdi_zscore,fuel-system_spfi_zscore
173,-1,65.0,19,4,102.4,175.6,66.5,54.9,2326,122,...,-0.376845,-0.072169,-0.072169,-0.245845,-0.704361,-0.330447,-0.072169,1.092329,-0.221163,-0.072169


In [282]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 193 entries, 0 to 204
Data columns (total 100 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   symboling                       193 non-null    int64  
 1   normalized-losses               193 non-null    float64
 2   make                            193 non-null    int64  
 3   num-of-doors                    193 non-null    int64  
 4   wheel-base                      193 non-null    float64
 5   length                          193 non-null    float64
 6   width                           193 non-null    float64
 7   height                          193 non-null    float64
 8   curb-weight                     193 non-null    int64  
 9   engine-size                     193 non-null    int64  
 10  bore                            193 non-null    float64
 11  stroke                          193 non-null    float64
 12  compression-ratio               193 non-

### Kết Quả

- Toàn bộ các cột được tạo ra từ quá trình chuyển đổi ở bước 03 đã được tạo z-score.
- Shape hiện tại là (193, 100), trông đó 193 dòng đều non-null.